staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter_test.go

```diff
@@ -0,0 +1,195 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conversion
+
+import (
+	"reflect"
+	"strings"
+	"testing"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/util/webhook"
+)
+
+func TestConversion(t *testing.T) {
+	tests := []struct {
+		Name            string
+		ValidVersions   []string
+		ClusterScoped   bool
+		ToVersion       string
+		SourceObject    runtime.Object
+		ExpectedObject  runtime.Object
+		ExpectedFailure string
+	}{
+		{
+			Name:          "simple_conversion",
+			ValidVersions: []string{"example.com/v1", "example.com/v2"},
+			ClusterScoped: false,
+			ToVersion:     "example.com/v2",
+			SourceObject: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "example.com/v1",
+					"other":      "data",
+					"kind":       "foo",
+				},
+			},
+			ExpectedObject: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "example.com/v2",
+					"other":      "data",
+					"kind":       "foo",
+				},
+			},
+			ExpectedFailure: "",
+		},
+		{
+			Name:          "failed_conversion_invalid_gv",
+			ValidVersions: []string{"example.com/v1", "example.com/v2"},
+			ClusterScoped: false,
+			ToVersion:     "example.com/v3",
+			SourceObject: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "example.com/v1",
+					"other":      "data",
+				},
+			},
+			ExpectedFailure: "invalid group/version: example.com/v3",
+		},
+		{
+			Name:          "simple_list_conversion",
+			ValidVersions: []string{"example.com/v1", "example.com/v2"},
+			ClusterScoped: false,
+			ToVersion:     "example.com/v2",
+			SourceObject: &unstructured.UnstructuredList{
+				Object: map[string]interface{}{
+					"apiVersion": "example.com/v1",
+					"kind":       "fooList",
+				},
+				Items: []unstructured.Unstructured{
+					{
+						Object: map[string]interface{}{
+							"apiVersion": "example.com/v1",
+							"kind":       "foo",
+							"other":      "data",
+						},
+					},
+					{
+						Object: map[string]interface{}{
+							"apiVersion": "example.com/v1",
+							"kind":       "foo",
+							"other":      "data2",
+						},
+					},
+				},
+			},
+			ExpectedObject: &unstructured.UnstructuredList{
+				Object: map[string]interface{}{
+					"apiVersion": "example.com/v2",
+					"kind":       "fooList",
+				},
+				Items: []unstructured.Unstructured{
+					{
+						Object: map[string]interface{}{
+							"apiVersion": "example.com/v2",
+							"kind":       "foo",
+							"other":      "data",
+						},
+					},
+					{
+						Object: map[string]interface{}{
+							"apiVersion": "example.com/v2",
+							"kind":       "foo",
+							"other":      "data2",
+						},
+					},
+				},
+			},
+			ExpectedFailure: "",
+		},
+		{
```

Comment:
do we need a test that passed in a list with `apiVersion: v1, kind: List`?

Reply:
do we do `v1.List` for CRDs? I remember we had a discussion that we don't do it for CRDs.

Reply:
you're right, found the thread: https://github.com/kubernetes/kubernetes/pull/67006#discussion_r232400766

probably worth hoisting the [result of that discussion](https://github.com/kubernetes/kubernetes/pull/67006#discussion_r232415232) into a comment in the code, just so I don't ask the same question in a month :)

Reply:
Added a comment to `crConverter.ConvertToVersion`